presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ApplyConnectorOptimization.java

```diff
@@ -0,0 +1,266 @@
+import com.facebook.presto.spi.ConnectorId;
+import com.facebook.presto.spi.ConnectorPlanOptimizer;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.plan.FilterNode;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.PlanNodeIdAllocator;
+import com.facebook.presto.spi.plan.TableScanNode;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class ApplyConnectorOptimization
+        implements PlanOptimizer
+{
+    static final Set<Class<? extends PlanNode>> CONNECTOR_ACCESSIBLE_PLAN_NODES = ImmutableSet.of(
+            FilterNode.class,
+            TableScanNode.class);
+
+    // for a leaf node that does not belong to any connector (e.g., ValuesNode)
+    private static final ConnectorId EMPTY_CONNECTOR_ID = new ConnectorId("$internal$" + ApplyConnectorOptimization.class + "_CONNECTOR");
+
+    private final Map<ConnectorId, Set<ConnectorPlanOptimizer>> connectorOptimizers;
+
+    public ApplyConnectorOptimization(Map<ConnectorId, Set<ConnectorPlanOptimizer>> connectorOptimizers)
+    {
+        this.connectorOptimizers = requireNonNull(connectorOptimizers, "connectorOptimizers is null");
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(types, "types is null");
+        requireNonNull(symbolAllocator, "symbolAllocator is null");
+        requireNonNull(idAllocator, "idAllocator is null");
+
+        if (connectorOptimizers.isEmpty()) {
+            return plan;
+        }
+
+        ConnectorSession connectorSession = session.toConnectorSession();
+
+        // retrieve all the connectors
+        ImmutableSet.Builder<ConnectorId> builder = ImmutableSet.builder();
+        getAllConnectorIds(plan, builder);
+
+        // only retain those having optimizers
+        Set<ConnectorId> connectorIds = new HashSet<>(builder.build());
+        connectorIds.retainAll(connectorOptimizers.keySet());
+        connectorIds = ImmutableSet.copyOf(connectorIds);
+        if (connectorIds.isEmpty()) {
+            // there is no connector optimizer applicable to the current plan
+            return plan;
```

Comment:
nit: I wonder if the code will be easier to read if this logic is replaced with a null check for optimizers inside the loop:

```
        Set<ConnectorId> connectorIds = builder.build();

        // for each connector, retrieve the set of subplans to optimize
        // TODO: what if a new connector is added by an existing one
        for (ConnectorId connectorId : connectorIds) {
            Set<ConnectorPlanOptimizer> optimizers = connectorOptimizers.get(connectorId);
            if (optimizers == null) {
                continue;
            }
```